crates/ruma-common/src/identifiers/macros.rs

```diff
@@ -24,8 +24,87 @@ macro_rules! partial_eq_string {
     }
 }
 
+macro_rules! owned_identifier {
+    ($owned:ident, $id:ident) => {
+        #[doc = concat!("Owned variant of ", stringify!($id))]
+        #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        pub struct $owned {
+            #[cfg(not(any(ruma_identifiers_storage = "Arc")))]
+            inner: Box<$id>,
+            #[cfg(ruma_identifiers_storage = "Arc")]
+            inner: std::sync::Arc<$id>,
+        }
+
+        impl AsRef<$id> for $owned {
+            fn as_ref(&self) -> &$id {
+                &self.inner
+            }
+        }
+
+        impl AsRef<str> for $owned {
+            fn as_ref(&self) -> &str {
+                &self.inner.as_ref()
+            }
+        }
+
+        impl std::ops::Deref for $owned {
+            type Target = $id;
+
+            fn deref(&self) -> &Self::Target {
+                &self.inner
+            }
+        }
+
+        impl std::borrow::Borrow<$id> for $owned {
+            fn borrow(&self) -> &$id {
+                self.as_ref()
+            }
+        }
+
+        impl From<&'_ $id> for $owned {
+            fn from(id: &$id) -> $owned {
+                $owned { inner: id.into() }
+            }
+        }
+
+        impl From<Box<$id>> for $owned {
+            fn from(b: Box<$id>) -> $owned {
+                Self {
+                    #[cfg(not(any(ruma_identifiers_storage = "Arc")))]
+                    inner: b,
+                    #[cfg(ruma_identifiers_storage = "Arc")]
+                    inner: std::sync::Arc::from(b),
```

Comment:
This should work, right? (plus I guess a `#[allow(clippy::whatever_lint_complains_about_identity_into)]`)

```suggestion
                    inner: b.into()
```

Reply:
If this works, I'd want to add the rustflags config into the CI matrix, so that this is continuously checked. 👍